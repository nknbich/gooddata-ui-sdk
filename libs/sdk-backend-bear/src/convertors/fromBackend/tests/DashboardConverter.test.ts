// (C) 2019-2021 GoodData Corporation
import * as uuid from "uuid";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";

import { convertDashboard } from "../DashboardConverter/index.js";
import {
    dashboardWithoutLayout,
    dashboardWithoutLayoutDependencies,
    exportFilterContext,
    emptyDashboard,
    exportTempFilterContext,
    dashboardWithExportTempFilterContextDependencies,
    emptyDashboardDependencies,
    dashboardWithFilterContext,
    dashboardWithFilterContextDependencies,
    dashboardWithExportFilterContextDependencies,
    visualizationClasses,
    dashboardWithLayout,
    dashboardWithLayoutDependencies,
    dashboardWithExtendedDateFilterConfig,
    dashboardWithLayoutAndCustomGridHeight,
} from "./DashboardConverter.fixtures.js";

describe("dashboard converter", () => {
    describe("convert dashboard", () => {
        beforeEach(() => {
            vi.spyOn(uuid, "v4").mockReturnValue("mocked-uuid");
        });

        afterEach(() => {
            vi.clearAllMocks();
        });

        it("should convert empty dashboard", () => {
            const convertedDashboard = convertDashboard(emptyDashboard, emptyDashboardDependencies);
            expect(convertedDashboard).toMatchSnapshot();
        });

        it("should convert dashboard with filter context", () => {
            const convertedDashboard = convertDashboard(
                dashboardWithFilterContext,
                dashboardWithFilterContextDependencies,
            );
            expect(convertedDashboard).toMatchSnapshot();
        });

        it("should override dashboard filter context with export filter context", () => {
            const convertedDashboard = convertDashboard(
                dashboardWithFilterContext,
                dashboardWithExportFilterContextDependencies,
                [],
                exportFilterContext.filterContext.meta.uri,
            );
            expect(convertedDashboard).toMatchSnapshot();
        });

        it("should convert dashboard with export temp filter context", () => {
            const convertedDashboard = convertDashboard(
                dashboardWithFilterContext,
                dashboardWithExportTempFilterContextDependencies,
                [],
                exportTempFilterContext.tempFilterContext.uri,
            );
            expect(convertedDashboard).toMatchSnapshot();
        });

        it("should convert dashboard without layout, and create implicit layout", () => {
            const convertedDashboard = convertDashboard(
                dashboardWithoutLayout,
                dashboardWithoutLayoutDependencies,
                visualizationClasses,
            );
            expect(convertedDashboard).toMatchSnapshot();
        });

        it("should convert dashboard with layout", () => {
            const convertedDashboard = convertDashboard(
                dashboardWithLayout,
                dashboardWithLayoutDependencies,
                visualizationClasses,
            );
            expect(convertedDashboard).toMatchSnapshot();
        });

        it("should convert dashboard with layout with custom gridHeight", () => {
            const convertedDashboard = convertDashboard(
                dashboardWithLayoutAndCustomGridHeight,
                dashboardWithLayoutDependencies,
                visualizationClasses,
            );
            expect(convertedDashboard).toMatchSnapshot();
        });

        it("should convert dashboard with extended date filters config", () => {
            const convertedDashboard = convertDashboard(
                dashboardWithExtendedDateFilterConfig,
                emptyDashboardDependencies,
            );
            expect(convertedDashboard).toMatchSnapshot();
        });
    });
});
