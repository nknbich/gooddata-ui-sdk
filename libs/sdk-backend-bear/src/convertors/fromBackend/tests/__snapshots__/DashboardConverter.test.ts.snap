// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dashboard converter > convert dashboard > should convert dashboard with export temp filter context 1`] = `
{
  "created": "2020-05-05 00:00:00",
  "createdBy": undefined,
  "dateFilterConfig": undefined,
  "description": "",
  "filterContext": {
    "created": "2020-05-05 00:00:00",
    "filters": [],
    "ref": {
      "uri": "/tempFilterContext",
    },
    "uri": "/tempFilterContext",
  },
  "identifier": "dashboardWithFilterContext",
  "isLocked": false,
  "layout": undefined,
  "plugins": undefined,
  "ref": {
    "uri": "/gdc/md/obj/dashboardWithFilterContext",
  },
  "shareStatus": "public",
  "tags": [],
  "title": "dashboardWithFilterContext",
  "type": "IDashboard",
  "updated": "2020-05-05 10:10:10",
  "updatedBy": undefined,
  "uri": "/gdc/md/obj/dashboardWithFilterContext",
}
`;

exports[`dashboard converter > convert dashboard > should convert dashboard with extended date filters config 1`] = `
{
  "created": "2020-05-05 00:00:00",
  "createdBy": undefined,
  "dateFilterConfig": {
    "addPresets": undefined,
    "filterName": "Filter",
    "hideGranularities": [
      "GDC.time.month",
    ],
    "hideOptions": undefined,
    "mode": "active",
  },
  "description": "",
  "filterContext": undefined,
  "identifier": "dashboardWithExtendedDateFilterConfig",
  "isLocked": false,
  "layout": undefined,
  "plugins": undefined,
  "ref": {
    "uri": "/gdc/md/obj/dashboardWithExtendedDateFilterConfig",
  },
  "shareStatus": "public",
  "tags": [],
  "title": "dashboardWithExtendedDateFilterConfig",
  "type": "IDashboard",
  "updated": "2020-05-05 10:10:10",
  "updatedBy": undefined,
  "uri": "/gdc/md/obj/dashboardWithExtendedDateFilterConfig",
}
`;

exports[`dashboard converter > convert dashboard > should convert dashboard with filter context 1`] = `
{
  "created": "2020-05-05 00:00:00",
  "createdBy": undefined,
  "dateFilterConfig": undefined,
  "description": "",
  "filterContext": {
    "description": "",
    "filters": [
      {
        "attributeFilter": {
          "attributeElements": {
            "uris": [
              "/test",
              "/test2",
            ],
          },
          "displayForm": {
            "uri": "/displayForm",
          },
          "filterElementsBy": [],
          "localIdentifier": "parent",
          "negativeSelection": false,
          "title": undefined,
        },
      },
      {
        "attributeFilter": {
          "attributeElements": {
            "uris": [
              "/test3",
              "/test4",
            ],
          },
          "displayForm": {
            "uri": "/displayForm34",
          },
          "filterElementsBy": [
            {
              "filterLocalIdentifier": "parent",
              "over": {
                "attributes": [
                  {
                    "uri": "attribute/uri",
                  },
                ],
              },
            },
          ],
          "localIdentifier": "child",
          "negativeSelection": false,
          "title": undefined,
        },
      },
      {
        "dateFilter": {
          "attribute": {
            "uri": "testAttr",
          },
          "from": -10,
          "granularity": "GDC.time.month",
          "to": 0,
          "type": "relative",
        },
      },
    ],
    "identifier": "filterContext",
    "ref": {
      "uri": "/gdc/md/obj/filterContext",
    },
    "title": "filterContext",
    "uri": "/gdc/md/obj/filterContext",
  },
  "identifier": "dashboardWithFilterContext",
  "isLocked": false,
  "layout": undefined,
  "plugins": undefined,
  "ref": {
    "uri": "/gdc/md/obj/dashboardWithFilterContext",
  },
  "shareStatus": "public",
  "tags": [],
  "title": "dashboardWithFilterContext",
  "type": "IDashboard",
  "updated": "2020-05-05 10:10:10",
  "updatedBy": undefined,
  "uri": "/gdc/md/obj/dashboardWithFilterContext",
}
`;

exports[`dashboard converter > convert dashboard > should convert dashboard with layout 1`] = `
{
  "created": "2020-05-05 00:00:00",
  "createdBy": undefined,
  "dateFilterConfig": undefined,
  "description": "",
  "filterContext": undefined,
  "identifier": "dashboardWithFilterContext",
  "isLocked": false,
  "layout": {
    "sections": [
      {
        "items": [
          {
            "size": {
              "xl": {
                "gridWidth": 12,
              },
            },
            "type": "IDashboardLayoutItem",
            "widget": {
              "dateDataSet": undefined,
              "description": "",
              "drills": [
                {
                  "localIdentifier": "mockeduuid",
                  "origin": {
                    "measure": {
                      "localIdentifier": "m1",
                    },
                    "type": "drillFromMeasure",
                  },
                  "target": {
                    "identifier": "dashboardId",
                  },
                  "transition": "in-place",
                  "type": "drillToDashboard",
                },
              ],
              "identifier": "visualizationWidgetHeadline",
              "ignoreDashboardFilters": [
                {
                  "displayForm": {
                    "uri": "/gdc/md/displayForm",
                  },
                  "type": "attributeFilterReference",
                },
              ],
              "insight": {
                "uri": "/gdc/md/obj/visualizationHeadline",
              },
              "properties": {},
              "ref": {
                "uri": "/gdc/md/obj/visualizationWidgetHeadline",
              },
              "title": "visualizationWidgetHeadline",
              "type": "insight",
              "uri": "/gdc/md/obj/visualizationWidgetHeadline",
            },
          },
          {
            "size": {
              "xl": {
                "gridWidth": 6,
              },
            },
            "type": "IDashboardLayoutItem",
            "widget": {
              "dateDataSet": undefined,
              "description": "",
              "drills": [
                {
                  "localIdentifier": "mockeduuid",
                  "origin": {
                    "measure": {
                      "localIdentifier": "m1",
                    },
                    "type": "drillFromMeasure",
                  },
                  "target": {
                    "uri": "/targetVisualization",
                  },
                  "transition": "pop-up",
                  "type": "drillToInsight",
                },
              ],
              "identifier": "visualizationWidgetBarChart",
              "ignoreDashboardFilters": [],
              "insight": {
                "uri": "/gdc/md/obj/visualizationBarChart",
              },
              "properties": {},
              "ref": {
                "uri": "/gdc/md/obj/visualizationWidgetBarChart",
              },
              "title": "visualizationWidgetBarChart",
              "type": "insight",
              "uri": "/gdc/md/obj/visualizationWidgetBarChart",
            },
          },
          {
            "size": {
              "xl": {
                "gridWidth": 2,
              },
            },
            "type": "IDashboardLayoutItem",
            "widget": {
              "dateDataSet": undefined,
              "description": "",
              "drills": [
                {
                  "origin": {
                    "measure": {
                      "uri": "/measure",
                    },
                    "type": "drillFromMeasure",
                  },
                  "tab": "testTab",
                  "target": {
                    "uri": "test",
                  },
                  "transition": "in-place",
                  "type": "drillToLegacyDashboard",
                },
              ],
              "identifier": "kpiWidget",
              "ignoreDashboardFilters": [],
              "kpi": {
                "comparisonDirection": "growIsGood",
                "comparisonType": "previousPeriod",
                "metric": {
                  "uri": "/measure",
                },
              },
              "ref": {
                "uri": "/gdc/md/obj/kpiWidget",
              },
              "title": "kpiWidget",
              "type": "kpi",
              "uri": "/gdc/md/obj/kpiWidget",
            },
          },
        ],
        "type": "IDashboardLayoutSection",
      },
    ],
    "type": "IDashboardLayout",
  },
  "plugins": undefined,
  "ref": {
    "uri": "/gdc/md/obj/dashboardWithFilterContext",
  },
  "shareStatus": "public",
  "tags": [],
  "title": "dashboardWithFilterContext",
  "type": "IDashboard",
  "updated": "2020-05-05 10:10:10",
  "updatedBy": undefined,
  "uri": "/gdc/md/obj/dashboardWithFilterContext",
}
`;

exports[`dashboard converter > convert dashboard > should convert dashboard with layout with custom gridHeight 1`] = `
{
  "created": "2020-05-05 00:00:00",
  "createdBy": undefined,
  "dateFilterConfig": undefined,
  "description": "",
  "filterContext": undefined,
  "identifier": "dashboardWithFilterContext",
  "isLocked": false,
  "layout": {
    "sections": [
      {
        "items": [
          {
            "size": {
              "xl": {
                "gridHeight": 12,
                "gridWidth": 12,
              },
            },
            "type": "IDashboardLayoutItem",
            "widget": {
              "dateDataSet": undefined,
              "description": "",
              "drills": [
                {
                  "localIdentifier": "mockeduuid",
                  "origin": {
                    "measure": {
                      "localIdentifier": "m1",
                    },
                    "type": "drillFromMeasure",
                  },
                  "target": {
                    "identifier": "dashboardId",
                  },
                  "transition": "in-place",
                  "type": "drillToDashboard",
                },
              ],
              "identifier": "visualizationWidgetHeadline",
              "ignoreDashboardFilters": [
                {
                  "displayForm": {
                    "uri": "/gdc/md/displayForm",
                  },
                  "type": "attributeFilterReference",
                },
              ],
              "insight": {
                "uri": "/gdc/md/obj/visualizationHeadline",
              },
              "properties": {},
              "ref": {
                "uri": "/gdc/md/obj/visualizationWidgetHeadline",
              },
              "title": "visualizationWidgetHeadline",
              "type": "insight",
              "uri": "/gdc/md/obj/visualizationWidgetHeadline",
            },
          },
          {
            "size": {
              "xl": {
                "gridHeight": 6,
                "gridWidth": 6,
              },
            },
            "type": "IDashboardLayoutItem",
            "widget": {
              "dateDataSet": undefined,
              "description": "",
              "drills": [
                {
                  "localIdentifier": "mockeduuid",
                  "origin": {
                    "measure": {
                      "localIdentifier": "m1",
                    },
                    "type": "drillFromMeasure",
                  },
                  "target": {
                    "uri": "/targetVisualization",
                  },
                  "transition": "pop-up",
                  "type": "drillToInsight",
                },
              ],
              "identifier": "visualizationWidgetBarChart",
              "ignoreDashboardFilters": [],
              "insight": {
                "uri": "/gdc/md/obj/visualizationBarChart",
              },
              "properties": {},
              "ref": {
                "uri": "/gdc/md/obj/visualizationWidgetBarChart",
              },
              "title": "visualizationWidgetBarChart",
              "type": "insight",
              "uri": "/gdc/md/obj/visualizationWidgetBarChart",
            },
          },
          {
            "size": {
              "xl": {
                "gridHeight": 6,
                "gridWidth": 2,
              },
            },
            "type": "IDashboardLayoutItem",
            "widget": {
              "dateDataSet": undefined,
              "description": "",
              "drills": [
                {
                  "origin": {
                    "measure": {
                      "uri": "/measure",
                    },
                    "type": "drillFromMeasure",
                  },
                  "tab": "testTab",
                  "target": {
                    "uri": "test",
                  },
                  "transition": "in-place",
                  "type": "drillToLegacyDashboard",
                },
              ],
              "identifier": "kpiWidget",
              "ignoreDashboardFilters": [],
              "kpi": {
                "comparisonDirection": "growIsGood",
                "comparisonType": "previousPeriod",
                "metric": {
                  "uri": "/measure",
                },
              },
              "ref": {
                "uri": "/gdc/md/obj/kpiWidget",
              },
              "title": "kpiWidget",
              "type": "kpi",
              "uri": "/gdc/md/obj/kpiWidget",
            },
          },
        ],
        "type": "IDashboardLayoutSection",
      },
    ],
    "type": "IDashboardLayout",
  },
  "plugins": undefined,
  "ref": {
    "uri": "/gdc/md/obj/dashboardWithFilterContext",
  },
  "shareStatus": "public",
  "tags": [],
  "title": "dashboardWithFilterContext",
  "type": "IDashboard",
  "updated": "2020-05-05 10:10:10",
  "updatedBy": undefined,
  "uri": "/gdc/md/obj/dashboardWithFilterContext",
}
`;

exports[`dashboard converter > convert dashboard > should convert dashboard without layout, and create implicit layout 1`] = `
{
  "created": "2020-05-05 00:00:00",
  "createdBy": undefined,
  "dateFilterConfig": undefined,
  "description": "",
  "filterContext": undefined,
  "identifier": "dashboardWithFilterContext",
  "isLocked": false,
  "layout": {
    "sections": [
      {
        "items": [
          {
            "size": {
              "xl": {
                "gridWidth": 12,
              },
            },
            "type": "IDashboardLayoutItem",
            "widget": {
              "dateDataSet": undefined,
              "description": "",
              "drills": [
                {
                  "localIdentifier": "mockeduuid",
                  "origin": {
                    "measure": {
                      "localIdentifier": "m1",
                    },
                    "type": "drillFromMeasure",
                  },
                  "target": {
                    "uri": "/targetVisualization",
                  },
                  "transition": "pop-up",
                  "type": "drillToInsight",
                },
              ],
              "identifier": "visualizationWidgetBarChart",
              "ignoreDashboardFilters": [],
              "insight": {
                "uri": "/gdc/md/obj/visualizationBarChart",
              },
              "properties": {},
              "ref": {
                "uri": "/gdc/md/obj/visualizationWidgetBarChart",
              },
              "title": "visualizationWidgetBarChart",
              "type": "insight",
              "uri": "/gdc/md/obj/visualizationWidgetBarChart",
            },
          },
          {
            "size": {
              "xl": {
                "gridWidth": 2,
              },
            },
            "type": "IDashboardLayoutItem",
            "widget": {
              "dateDataSet": undefined,
              "description": "",
              "drills": [
                {
                  "origin": {
                    "measure": {
                      "uri": "/measure",
                    },
                    "type": "drillFromMeasure",
                  },
                  "tab": "testTab",
                  "target": {
                    "uri": "test",
                  },
                  "transition": "in-place",
                  "type": "drillToLegacyDashboard",
                },
              ],
              "identifier": "kpiWidget",
              "ignoreDashboardFilters": [],
              "kpi": {
                "comparisonDirection": "growIsGood",
                "comparisonType": "previousPeriod",
                "metric": {
                  "uri": "/measure",
                },
              },
              "ref": {
                "uri": "/gdc/md/obj/kpiWidget",
              },
              "title": "kpiWidget",
              "type": "kpi",
              "uri": "/gdc/md/obj/kpiWidget",
            },
          },
          {
            "size": {
              "xl": {
                "gridWidth": 2,
              },
            },
            "type": "IDashboardLayoutItem",
            "widget": {
              "dateDataSet": undefined,
              "description": "",
              "drills": [
                {
                  "localIdentifier": "mockeduuid",
                  "origin": {
                    "measure": {
                      "localIdentifier": "m1",
                    },
                    "type": "drillFromMeasure",
                  },
                  "target": {
                    "identifier": "dashboardId",
                  },
                  "transition": "in-place",
                  "type": "drillToDashboard",
                },
              ],
              "identifier": "visualizationWidgetHeadline",
              "ignoreDashboardFilters": [
                {
                  "displayForm": {
                    "uri": "/gdc/md/displayForm",
                  },
                  "type": "attributeFilterReference",
                },
              ],
              "insight": {
                "uri": "/gdc/md/obj/visualizationHeadline",
              },
              "properties": {},
              "ref": {
                "uri": "/gdc/md/obj/visualizationWidgetHeadline",
              },
              "title": "visualizationWidgetHeadline",
              "type": "insight",
              "uri": "/gdc/md/obj/visualizationWidgetHeadline",
            },
          },
        ],
        "type": "IDashboardLayoutSection",
      },
    ],
    "type": "IDashboardLayout",
  },
  "plugins": undefined,
  "ref": {
    "uri": "/gdc/md/obj/dashboardWithFilterContext",
  },
  "shareStatus": "public",
  "tags": [],
  "title": "dashboardWithFilterContext",
  "type": "IDashboard",
  "updated": "2020-05-05 10:10:10",
  "updatedBy": undefined,
  "uri": "/gdc/md/obj/dashboardWithFilterContext",
}
`;

exports[`dashboard converter > convert dashboard > should convert empty dashboard 1`] = `
{
  "created": "2020-05-05 00:00:00",
  "createdBy": undefined,
  "dateFilterConfig": undefined,
  "description": "",
  "filterContext": undefined,
  "identifier": "emptyDashboard",
  "isLocked": false,
  "layout": undefined,
  "plugins": undefined,
  "ref": {
    "uri": "/gdc/md/obj/emptyDashboard",
  },
  "shareStatus": "public",
  "tags": [],
  "title": "emptyDashboard",
  "type": "IDashboard",
  "updated": "2020-05-05 10:10:10",
  "updatedBy": undefined,
  "uri": "/gdc/md/obj/emptyDashboard",
}
`;

exports[`dashboard converter > convert dashboard > should override dashboard filter context with export filter context 1`] = `
{
  "created": "2020-05-05 00:00:00",
  "createdBy": undefined,
  "dateFilterConfig": undefined,
  "description": "",
  "filterContext": {
    "description": "",
    "filters": [],
    "identifier": "exportFilterContext",
    "ref": {
      "uri": "/gdc/md/obj/exportFilterContext",
    },
    "title": "exportFilterContext",
    "uri": "/gdc/md/obj/exportFilterContext",
  },
  "identifier": "dashboardWithFilterContext",
  "isLocked": false,
  "layout": undefined,
  "plugins": undefined,
  "ref": {
    "uri": "/gdc/md/obj/dashboardWithFilterContext",
  },
  "shareStatus": "public",
  "tags": [],
  "title": "dashboardWithFilterContext",
  "type": "IDashboard",
  "updated": "2020-05-05 10:10:10",
  "updatedBy": undefined,
  "uri": "/gdc/md/obj/dashboardWithFilterContext",
}
`;
